lib/os/onoff.c

```diff
@@ -390,128 +408,151 @@ int onoff_release(struct onoff_service *srv,
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+
+	if (get_client_type(cli) != ONOFF_CLIENT_INACTIVE) {
+		rv = -EINVAL;
+		goto out;
+	}
 
-	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) == 0) {
+	if ((mgr->flags & ONOFF_HAS_ERROR) == 0) {
 		rv = -EALREADY;
 		goto out;
 	}
 
-	if ((srv->flags & SERVICE_STATE_TRANSITION) == 0) {
+	if ((mgr->flags & ST_TRANSITION) == 0) {
 		reset = true;
-		srv->flags |= SERVICE_STATE_TRANSITION;
+		mgr->flags |= ST_TRANSITION;
 	}
 
 out:
 	if (rv >= 0) {
-		sys_slist_append(&srv->clients, &cli->node);
+		set_client_type(cli, ONOFF_CLIENT_RESET);
+		sys_slist_append(&mgr->clients, &cli->node);
 	}
 
-	k_spin_unlock(&srv->lock, key);
+	k_spin_unlock(&mgr->lock, key);
 
 	if (reset) {
-		srv->reset(srv, onoff_reset_notify);
+		mgr->transitions->reset(mgr, onoff_reset_notify);
 	}
 
 	return rv;
 }
 
-int onoff_cancel(struct onoff_service *srv,
+int onoff_cancel(struct onoff_manager *mgr,
 		 struct onoff_client *cli)
 {
-	int rv = validate_args(srv, cli);
+	int rv = validate_args(mgr, cli);
 
 	if (rv < 0) {
 		return rv;
 	}
 
 	rv = -EALREADY;
-	k_spinlock_key_t key = k_spin_lock(&srv->lock);
-	u32_t state = srv->flags & SERVICE_STATE_MASK;
-
-	/* Can't remove the last client waiting for the in-progress
-	 * transition, as there would be nobody to receive the
-	 * completion notification, which might indicate a service
-	 * error.
-	 */
-	if (sys_slist_find_and_remove(&srv->clients, &cli->node)) {
-		rv = 0;
-		if (sys_slist_is_empty(&srv->clients)
-		    && (state != SERVICE_STATE_TO_OFF)) {
-			rv = -EWOULDBLOCK;
-			sys_slist_append(&srv->clients, &cli->node);
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+	u32_t state = mgr->flags & ST_MASK;
+
+	if (sys_slist_find_and_remove(&mgr->clients, &cli->node)) {
+		switch (get_client_type(cli)) {
+		case ONOFF_CLIENT_REQUEST:
+			/* Requests can be cancelled whether transitioning to on or off.
+			 * They're cancelled immediately.
+			 */
+			__ASSERT_NO_MSG((mgr->flags & ONOFF_HAS_ERROR) == 0);
+			__ASSERT_NO_MSG((state == ST_TO_ON)
+					|| (state == ST_TO_OFF));
+			rv = 0;
+			break;
+		case ONOFF_CLIENT_RELEASE:
+			__ASSERT_NO_MSG((mgr->flags & ONOFF_HAS_ERROR) == 0);
+			__ASSERT_NO_MSG(state == ST_TO_OFF);
+
+			/* A release can only be present when
+			 * transitioning to off, and when cancelled it
+			 * converts to a request that goes back on the
+			 * client list to be completed later.
+			 */
+			set_client_type(cli, ONOFF_CLIENT_REQUEST);
+			mgr->refs -= 1;
+			rv = 1;
+			sys_slist_append(&mgr->clients, &cli->node);
+			break;
+		case ONOFF_CLIENT_RESET:
+			__ASSERT_NO_MSG((mgr->flags & ONOFF_HAS_ERROR) != 0);
+			rv = 0;
+			break;
+		default:
+			break;
 		}
-	} else if (srv->releaser == cli) {
-		/* must be waiting for TO_OFF to complete */
-		rv = -EWOULDBLOCK;
 	}
 
-	k_spin_unlock(&srv->lock, key);
+	k_spin_unlock(&mgr->lock, key);
 
 	if (rv == 0) {
-		notify_one(srv, cli, -ECANCELED);
+		notify_one(mgr, cli, -ECANCELED);
```

Comment:
this notification is now somehow artificial and redundant. If cancellation returning 0 tells user that request/release is cancelled. No need to handle that again in the callback. Also that reduces a chance of error code conflict (service returning -ECANCELED error).

Reply:
That was always intended so there would be exactly one place in the code to process the completion that confirms that the client structure has been released.  Not completing might make a client more complex.  I agree about deconflicting with the service error space, though, so this will probably change.